mypy/checkexpr.py

```diff
@@ -3108,112 +3108,134 @@ def infer_lambda_type_using_context(self, e: LambdaExpr) -> Tuple[Optional[Calla
-                        return AnyType(TypeOfAny.special_form)
-                    if not self.chk.in_checked_function():
-                        return AnyType(TypeOfAny.unannotated)
-                    if self.chk.scope.active_class() is not None:
-                        self.chk.fail(message_registry.SUPER_OUTSIDE_OF_METHOD_NOT_SUPPORTED, e)
-                        return AnyType(TypeOfAny.from_error)
-                    method = self.chk.scope.top_function()
-                    assert method is not None
-                    args = method.arguments
-                    # super() in a function with empty args is an error; we
-                    # need something in declared_self.
-                    if not args:
-                        self.chk.fail(message_registry.SUPER_ENCLOSING_POSITIONAL_ARGS_REQUIRED, e)
-                        return AnyType(TypeOfAny.from_error)
-                    declared_self = args[0].variable.type or fill_typevars(e.info)
-                    return analyze_member_access(name=e.name,
-                                                 typ=fill_typevars(e.info),
-                                                 is_lvalue=False,
-                                                 is_super=True,
-                                                 is_operator=False,
-                                                 original_type=declared_self,
-                                                 override_info=base,
-                                                 context=e,
-                                                 msg=self.msg,
-                                                 chk=self.chk,
-                                                 in_literal_context=self.is_literal_context())
-            assert False, 'unreachable'
         else:
-            # Invalid super. This has been reported by the semantic analyzer.
+            self.chk.fail(message_registry.TOO_MANY_ARGS_FOR_SUPER, e)
             return AnyType(TypeOfAny.from_error)
 
+        # Imprecisely assume that the type is the current class
+        if (isinstance(type_type, AnyType)
+                or (isinstance(type_type, TypeType) and isinstance(type_type.item, AnyType))):
+            type_type = current_type
+
+        if (not isinstance(type_type, TypeType)
+                and not (isinstance(type_type, FunctionLike) and type_type.is_type_obj())):
+            self.msg.first_argument_for_super_must_be_type(type_type, e)
+            return AnyType(TypeOfAny.from_error)
+
+        # Imprecisely assume that the instance is of the current class
+        if isinstance(instance_type, AnyType):
+            instance_type = current_instance
+        elif isinstance(instance_type, TypeType) and isinstance(instance_type.item, AnyType):
+            instance_type = current_type
+
+        return type_type, instance_type
+
+    def _type_info(self, typ: Type) -> Optional[TypeInfo]:
```

Comment:
Style nit: since this doesn't use `self` in the body, this could be changed to a module-level function. If you do this, the name would have to be changed, since we don't generally use `_` prefix at module level. Suggestion: `type_info_from_type`.